File: app/src/main/java/com/relic/data/PostRepositoryImpl.java
package com.relic.data;


import android.arch.lifecycle.LiveData;
import android.content.Context;
import android.os.AsyncTask;
import android.util.Log;

import com.android.volley.Request;
import com.android.volley.RequestQueue;
import com.android.volley.Response;
import com.android.volley.VolleyError;
import com.android.volley.toolbox.StringRequest;
import com.android.volley.toolbox.Volley;
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.relic.R;
import com.relic.data.dao.PostDao;
import com.relic.data.entities.PostEntity;
import com.relic.data.models.PostListingModel;
import com.relic.data.models.PostModel;
import com.relic.domain.Post;
import com.relic.domain.post.PostImpl;

import org.json.simple.JSONArray;
import org.json.simple.JSONObject;
import org.json.simple.parser.JSONParser;
import org.json.simple.parser.ParseException;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Map;

public class PostRepositoryImpl implements PostRepository {
  private final String ENDPOINT = "https://oauth.reddit.com/r/";
  private final String userAgent = "android:com.relic.Relic (by /u/boiledbuns)";
  private final String TAG = "POST_REPO";

  private Context context;
  private String authToken;
  private JSONParser JSONParser;
  private ApplicationDB appDB;

  RequestQueue requestQueue;


  public PostRepositoryImpl(Context context) {
    this.context = context;
    requestQueue = Volley.newRequestQueue(context);
    JSONParser = new JSONParser();

    // get the oauth token from the app's shared preferences
    String authKey = context.getResources().getString(R.string.AUTH_PREF);
    String tokenKey = context.getResources().getString(R.string.TOKEN_KEY);
    authToken = context.getSharedPreferences(authKey, Context.MODE_PRIVATE)
        .getString(tokenKey, "DEFAULT");

    // initialize reference to the database
    appDB = ApplicationDB.getDatabase(context);
  }


  public LiveData<List<PostModel>> getPosts(String subreddit) {
    retrieveMorePosts(subreddit);
    return appDB.getPostDao().getSubredditPosts("1");
  }


  @Override
  public void retrieveMorePosts(String subreddit) {
    // create the new request and submit it
    requestQueue.add(new RedditOauthRequest(Request.Method.GET, ENDPOINT + subreddit,
        new Response.Listener<String>() {
          @Override
          public void onResponse(String response) {
            //Log.d(TAG, response);
            try {
              parsePosts(response);
            }
            catch (ParseException error) {
              Log.d(TAG, "Error: " + error.getMessage());
            }
          }
        }, new Response.ErrorListener() {
      @Override
      public void onErrorResponse(VolleyError error) {
        Log.d(TAG, "Error: " + error.networkResponse.headers.toString());
      }
    }
    ));
  }


  /**
   * Parses the response from the api and stores the data in the persistence layer
   * @param response the json response from the server with the listing object
   * @throws ParseException
   */
  private void parsePosts(String response) throws ParseException {
    JSONObject listingData = (JSONObject) ((JSONObject) JSONParser.parse(response)).get("data");
    JSONArray listingPosts = (JSONArray) (listingData).get("children");

    Iterator postIterator = listingPosts.iterator();
    List <PostEntity> postEntities = new ArrayList<>();

    // GSON reader to unmarshall the json response
    Gson gson = new GsonBuilder().create();
    Log.d(TAG, "dank = " + response);

    // generate the list of posts using the json array
    while (postIterator.hasNext()) {
      JSONObject post = (JSONObject) ((JSONObject) postIterator.next()).get("data");

//      for (int i = 0; i < Math.ceil(post.toJSONString().length()/900); i ++) {
//        Log.d(TAG + " " + i, post.toJSONString().substring(0 + i*900, 900 + i*900));
//      }

      // demarshall the object and add it into a list
      //Log.d(TAG, "post : " + post.get("title") + " "+ post.get("edited"));
      postEntities.add(gson.fromJson(post.toJSONString(), PostEntity.class));

      //Log.d(TAG, "post keys " + post.keySet().toString());
    }

    // insert all the post entities into the db
    //Log.d(TAG, postEntities.size() + " posts retrieved");
    new InsertPostsTask(appDB, postEntities).execute();
  }


  class RedditOauthRequest extends StringRequest {
    private RedditOauthRequest(int method, String url, Response.Listener<String> listener,
                              Response.ErrorListener errorListener) {
      super(method, url, listener, errorListener);
    }

    public Map<String, String> getHeaders() {
      Map <String, String> headers = new HashMap<>();

      // generate the credential string for oauth
      String credentials = "bearer " + authToken;
      headers.put("Authorization", credentials);
      headers.put("User-Agent", userAgent);

      return headers;
    }
  }

  static class InsertPostsTask extends AsyncTask<String, Integer, Integer> {
    ApplicationDB appDB;
    List<PostEntity> postList;

    InsertPostsTask(ApplicationDB db, List<PostEntity> posts) {
      super();
      appDB = db;
      postList = posts;
    }

    @Override
    protected Integer doInBackground(String... strings) {
      appDB.getPostDao().insertPosts(postList);
      return null;
    }
  }

}

File: app/src/main/java/com/relic/data/PostRepositoryImpl.kt
package com.relic.data

import android.arch.lifecycle.LiveData
import android.arch.lifecycle.MediatorLiveData
import android.arch.lifecycle.MutableLiveData
import android.content.Context
import android.os.AsyncTask
import android.util.Log
import com.android.volley.NoConnectionError

import com.relic.R
import com.relic.data.deserializer.ParsedPostsData
import com.relic.data.deserializer.PostDeserializerImpl
import com.relic.network.NetworkRequestManager
import com.relic.data.gateway.PostGateway
import com.relic.data.gateway.PostGatewayImpl
import com.relic.network.request.RelicOAuthRequest
import com.relic.presentation.callbacks.RetrieveNextListingCallback
import com.relic.data.entities.ListingEntity
import com.relic.data.entities.PostEntity
import com.relic.data.entities.PostSourceEntity
import com.relic.data.models.PostModel
import com.relic.data.repository.RepoConstants.ENDPOINT
import com.relic.network.request.RelicRequestError
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

import org.json.simple.parser.ParseException
import java.lang.Exception

import java.util.ArrayList

import javax.inject.Inject

/**
 * This repository is used for accessing posts either by:
 * a) getting a livedata reference for locally stored posts
 * b) retrieving posts from the network and storing them locally
 *
 * Note: Cancellation of parent coroutines automatically propagate downward to
 * all its children. When performing retrieval methods (eg. post retrieval) from
 * the network, the process should not be cancelled even if its parent coroutine
 * is cancelled.
 *
 * I initially designed the retrieval methods as non-suspending functions that internally
 * launched a coroutine from the Global scope but realized suspend functions offered more
 * benefits:
 * a) The method is poorly thought out and makes cancellation more difficult than necessary
 * b) Because the new methods are now suspending, the caller has control over the coroutine
 * scope. This is certainly more flexible as there are some (though rare) cases where we can
 * supply a non-global scope that we can close at our own convenience to cancel retrieval
 *
 *  As a result, the retrieval methods are now suspend functions
 */
class PostRepositoryImpl @Inject constructor(
    appContext: Context,
    private val requestManager: NetworkRequestManager
) : PostRepository {
    private val TAG = "POST_REPO"

    companion object {
        // keys for the "after" value for listings
        private const val KEY_FRONTPAGE = "frontpage"
        private const val KEY_ALL = "all"
        private const val KEY_OTHER = "other"
    }

    private val sortTypesWithScope = arrayOf(
        PostRepository.SortType.HOT,
        PostRepository.SortType.RISING,
        PostRepository.SortType.TOP
    )

    private val appDB: ApplicationDB = ApplicationDB.getDatabase(appContext)
    // TODO convert this to DI
    private val postDeserializer = PostDeserializerImpl(appContext)

    override val postGateway: PostGateway = PostGatewayImpl(appContext, requestManager)

    // region interface methods

    override fun getPosts(postSource: PostRepository.PostSource) : LiveData<List<PostModel>> {
        return when (postSource) {
            is PostRepository.PostSource.Subreddit -> appDB.postDao.getPostsFromSubreddit(postSource.subredditName)
            is PostRepository.PostSource.Frontpage -> appDB.postDao.getPostsFromFrontpage()
            is PostRepository.PostSource.User -> {
                when (postSource.retrievalOption) {
                    PostRepository.RetrievalOption.Submitted -> appDB.userPostingDao.getUserPosts()
                    PostRepository.RetrievalOption.Comments -> MutableLiveData()
                    PostRepository.RetrievalOption.Saved -> appDB.userPostingDao.getUserSavedPosts()
                    PostRepository.RetrievalOption.Upvoted -> appDB.userPostingDao.getUserUpvotedPosts()
                    PostRepository.RetrievalOption.Downvoted -> appDB.userPostingDao.getUserDownvotedPosts()
                    PostRepository.RetrievalOption.Gilded -> appDB.userPostingDao.getUserGilded()
                    PostRepository.RetrievalOption.Hidden -> appDB.userPostingDao.getUserHidden()
                }
            }
            else -> appDB.postDao.getPostsFromAll()
        }
    }

    override suspend fun retrieveMorePosts(
        postSource: PostRepository.PostSource,
        listingAfter: String
    ) {
        // change the api endpoint to access the next post listing
        val ending = when (postSource) {
            is PostRepository.PostSource.Subreddit -> "r/${postSource.subredditName}"
            is PostRepository.PostSource.User -> "user/${postSource.username}/${postSource.retrievalOption.name.toLowerCase()}"
            else -> ""
        }

        Log.d(TAG, "retrieve more posts : api url : $ENDPOINT$ending?after=$listingAfter")
        try {
            val response = requestManager.processRequest(
                method = RelicOAuthRequest.GET,
                url = "$ENDPOINT$ending?after=$listingAfter"
            )
            Log.d(TAG, "more posts $response")

            val listingKey = getListingKey(postSource)
            val parsedData = postDeserializer.parsePosts(response, postSource, listingKey)

            insertParsedPosts(parsedData)
        } catch (e: Exception) {
            throw DomainTransfer.handleException("retrieve more posts", e) ?: e
        }

    }

    override suspend fun getNextPostingVal(callback: RetrieveNextListingCallback, postSource: PostRepository.PostSource) {
        val key = getListingKey(postSource)

        withContext(Dispatchers.IO) {
            // get the "after" value for the most current sub listing
            val subAfter = appDB.listingDAO.getNext(key)
            callback.onNextListing(subAfter)
        }
    }

    override fun getPost(postFullName: String): LiveData<PostModel> {
        return appDB.postDao.getSinglePost(postFullName)
    }

    @Throws(RelicRequestError::class)
    override suspend fun retrieveSortedPosts(
        postSource: PostRepository.PostSource,
        sortType: PostRepository.SortType,
        sortScope: PostRepository.SortScope
    ) {
        // convert into a builder to make it easier to build api url
        // generate the ending of the request url based on the source type
        var ending = ENDPOINT + when (postSource) {
            is PostRepository.PostSource.Subreddit -> "r/${postSource.subredditName}"
            is PostRepository.PostSource.User -> {
                "user/${postSource.username}/${postSource.retrievalOption.name.toLowerCase()}?sort=${sortType.name.toLowerCase()}&t=${sortScope.name.toLowerCase()}"
            }
            else -> ""
        }

        // modify the endpoint based on the sorting options selected by the user
        if (sortType != PostRepository.SortType.DEFAULT && postSource !is PostRepository.PostSource.User) {
            // build the appropriate endpoint based on the "sort by" code and time scope
            ending += "/${sortType.name.toLowerCase()}/"

            // only add sort scope for the options that accept it
            if (sortTypesWithScope.contains(sortType)) ending += "?t=" + sortScope.name.toLowerCase()
        }

        Log.d(TAG, "retrieve sorted posts api url : $ending")
        coroutineScope {
            try {
                val response = requestManager.processRequest(
                    method = RelicOAuthRequest.GET,
                    url = ending
                )
                val clear = launch { clearAllPostsFromSource(postSource) }
                Log.d(TAG, "retrieve posts response :  $response")

                val listingKey = getListingKey(postSource)
                val parsedData = postDeserializer.parsePosts(response, postSource, listingKey)
                Log.d(TAG, "retrieve more posts : after ${parsedData.listingEntity.afterPosting}")

                clear.join()
                insertParsedPosts(parsedData)
            } catch (e: Exception) {
                throw DomainTransfer.handleException("retrieve account", e) ?: e
            }
        }
    }

    override suspend fun retrievePost(
        subredditName: String,
        postFullName: String,
        postSource: PostRepository.PostSource
    ) {
        val ending = "r/$subredditName/comments/${postFullName.substring(3)}"

        try {
            val response = requestManager.processRequest(
                method = RelicOAuthRequest.GET,
                url = ENDPOINT + ending
            )

            postDeserializer.parsePost(response).apply {
                postEntity.visited = true

                withContext(Dispatchers.IO) {
                    appDB.postDao.insertPost(postEntity)
                    appDB.postSourceDao.insertPostSources(listOf(postSourceEntity))
                }
            }

        } catch (e: Exception) {
            throw DomainTransfer.handleException("retrieve account", e) ?: e
        }
    }

    override suspend fun clearAllPostsFromSource(postSource: PostRepository.PostSource) {
        withContext(Dispatchers.IO) {
            when (postSource) {
                is PostRepository.PostSource.Frontpage -> appDB.postSourceDao.removeAllFrontpageAsSource()
                is PostRepository.PostSource.All -> appDB.postSourceDao.removeAllAllAsSource()
                is PostRepository.PostSource.Subreddit -> appDB.postSourceDao.removeAllSubredditAsSource(postSource.subredditName)
                is PostRepository.PostSource.User -> {
                    appDB.postSourceDao.apply {
                        when (postSource.retrievalOption) {
                            PostRepository.RetrievalOption.Submitted -> removeAllUserSubmittedAsSource()
                            PostRepository.RetrievalOption.Comments -> removeAllUserCommentsAsSource()
                            PostRepository.RetrievalOption.Saved -> removeAllUserSavedAsSource()
                            PostRepository.RetrievalOption.Upvoted -> removeAllUserUpvotedAsSource()
                            PostRepository.RetrievalOption.Downvoted -> removeAllUserDownvotedAsSource()
                            PostRepository.RetrievalOption.Gilded -> removeAllUserGildedAsSource()
                            PostRepository.RetrievalOption.Hidden -> removeAllUserHiddenAsSource()
                        }
                    }
                }
            }

            // remove all the source entities that no longer correspond to any remaining posts
            appDB.postSourceDao.removeAllUnusedSources()
        }
    }

    // endregion interface methods


    private suspend fun insertParsedPosts(parsedPosts : ParsedPostsData) {
        withContext(Dispatchers.IO) {
            parsedPosts.apply {
                if (postEntities.isNotEmpty()) appDB.postDao.insertPosts(postEntities)
                if (commentEntities.isNotEmpty()) appDB.commentDAO.insertComments(commentEntities)

                appDB.postSourceDao.insertPostSources(postSourceEntities)
                appDB.listingDAO.insertListing(listingEntity)
            }
        }
    }

    /**
     * builds the key used for retrieving the "after" value for a listing using its associated
     * post source
     */
    private fun getListingKey(postSource : PostRepository.PostSource) : String {
        return when (postSource) {
            is PostRepository.PostSource.Subreddit -> postSource.subredditName
            is PostRepository.PostSource.Frontpage -> KEY_FRONTPAGE
            is PostRepository.PostSource.All -> KEY_ALL
            is PostRepository.PostSource.User -> postSource.username + postSource.retrievalOption.name
            else -> KEY_OTHER
        }
    }
}
